### PR TITLE
[AIRFLOW-4517] allow for airflow config to influence flask configuration

### PR DESCRIPTION
**Pull request by mb-m**
Add the ability to use a [flask_web] section to set flask configuration
variables. This is particularly useful with the introduction of origin=
on some of the dags.html page links (trigger, particularly), as without
it, there's no way of specifying https in the url_for() flask function.
Right now, without a good framework for seeing the effect of flask vars,
it's difficult to write a perfect test, but I've added something that showed
me it was there.

### CODE REVIEW

**Comment by mb-m** [comment]
The CI failure here appears to have nothing to do with my patch 😢

**Comment by potiuk** [comment]
@mb-m -> are you sure you want to merge it to v1-10 branch ? I think you should work from master instead. It does not seem your change is specific to 1.10.x releases.

Usually the process is that you make PR to merge to master (2.0.0-dev) and then committers cherry-pick it to the v1-10-test branch. The v1-10-test branch occasionally fails because it is used to test Py27 compatibility of cherry-picked changes and that way it is usually only. This is what happened with this build. That's why v1-10 branch is pretty much only used by committers.

So in short: if you want this change to be also in future versions of airflow, please rebase it and request it to be merged to master.

**Comment by mb-m** [comment]
@potiuk In this case, the patch is against v1.10.x, as I found the issue due to changes introduced in 1.10.3 - I don't believe the origin= is present on `master` when I looked, as the dags page for master is a little different. It's plausible we should also do this on master, however.

1.10.2 -> 1.10.3 has some issues with manual triggers which we're currently using while working out how our airflow is going to work, and it'd be nice if there's a 1.10.4 release to fix these - what's the plan between 1.10.x (x > 3) and 2.0.0?

**Comment by mb-m** [comment]
@potiuk the above notwithstanding, I'm happy to listen to any further advice you have (sorry if that's not clear in my message).

**Comment by potiuk** [comment]
Thenm if the bug has different fixes for 1.10 and 2.0 and they are significantly different, then ideally we should have two parallel PRs. We knew that the time is close when merging master -> v1.10x will become too painful and for a while I think we will have to live with this dualism in place. Most of the changes are easy to apply master-> 1.10 still.

The current general roadmap for 2.0 is being created in https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+2.0 . Those particular issue you mention - if this is a JIRA issue and it will be solved in 2.0 we will try to backport it to 1.10.x. It's all issue-by-issue case and fairly 'distributed' so there is no central place to see what's implemented.

**Comment by potiuk** [comment]
And BTW. the v1-10-test branch now builds fine, so you can rebase.

**Comment by mb-m** [comment]
@potiuk I can try and come up with a PR for master - it's going to be a bit more difficult, as we don't have a 2.x-pre (master) airflow deploy anywhere to see if this works, it may be that this applies cleanly, of course 😄 and that might be enough.

File: tests/cli/test_cli.py [hunk 1]
```diff
@@ -188,6 +188,20 @@ def test_cli_rbac_webserver_debug(self):
         p.terminate()
         p.wait()
 
+    def test_cli_webserver_flask_debug(self):
+        env = os.environ.copy()
+        env['AIRFLOW__WEBSERVER__RBAC'] = 'True'
+        env['AIRFLOW__FLASK_WEB__PREFERRED_URL_SCHEME'] = 'https'
+        p = psutil.Popen(["airflow", "webserver", "-d"], env=env)
```
Comment (ashb) [inline tests/cli/test_cli.py:195]: This test doesn't really assert that we've configured the webserver right. Better would be to use mocking

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -857,21 +857,31 @@ def webserver(args):
...
+    flask_web_config = conf.getsection('flask_web')
+    if flask_web_config:
+        flask_web_upper_update = {}
+        for k in flask_web_config:
+            K = k.upper()
+            if k != K:
+                flask_web_upper_update[k] = K
+        for k in flask_web_upper_update:
+            flask_web_config[flask_web_upper_update[k]] = flask_web_config[k]
+            del flask_web_config[k]
```
Comment (ashb) [inline airflow/bin/cli.py]: This whole block is a little bit hard to follow. What is it doing? What is the purpose.

I suspect that https://github.com/apache/airflow/pull/5379 might make much of this un-needed too.

Comment (ashb) [inline airflow/bin/cli.py:869]: This whole block is a little bit hard to follow. What is it doing? What is the purpose.

I suspect that https://github.com/apache/airflow/pull/5379 might make much of this un-needed too.

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by ashb** [comment]
You should also create a PR that targets against the master branch, not directly on to the release branch please.

**Comment by mb-m** [comment]
@ashb as discussed above, `master` doesn't appear to have the same structure or issues.

File: airflow/bin/cli.py (hunk 2 above)
Comment (mb-m) [reply airflow/bin/cli.py:869]: The flask config names are upper-cased (http://flask.pocoo.org/docs/1.0/config), but the airflow ones are lower-cased. When you get variables set by environment (same as many other airflow configs), you want these to operate in the correct way too. Unfortunately the handler for environment in config has already lower-cased them by this point.

**Comment by ashb** [comment]
Master will have a slightly different code structure, yes. (the "RBAC" part has become the only part on master) Please target against that version and the committers will deal with back-porting this.

**Comment by mb-m** [comment]
@ashb fair enough, I'll close this PR, and come back to this at some point later.
